[flutter_plugin_tools] Add a command to lint Android code

stuartmorgan opened the pull request:
Adds a new `lint-android` command to run `gradlew lint` on Android plugins.

Also standardizes the names of the Cirrus tasks that run all the build and platform-specific (i.e., not Dart unit test) tests for each platform, as they were getting unnecessarily long and complex in some cases.

Fixes https://github.com/flutter/flutter/issues/87071

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I [updated pubspec.yaml](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates) with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [ ] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code

stuartmorgan commented:
This is a WIP since the new command will fail on many plugins. @blasten How do you want to proceed here:
1. `--exclude` all plugins that fail, and enable them as they are made to work?
2. Configure all plugins to not fail on error for now via the XML configuration, and remove that as they are fixed up?
3. Not add it to CI at all yet, get things working locally, then turn it on later?
4. ???

My preference is (1) since it uses the established place for tracking what's currently not turned on, but I can see a good argument for (2) as well given that we'd have output on the bots to work from right away. (We'll need an issues tracking getting it fully running for whatever option we choose, of course.)

Other open questions:
- Are there baseline options you want to set for all plugins before turning this on, or do you want to just make adjusting that part of fully opting each plugin into the run?
- Are you fine with punting on the XML parser for now, and just showing the partial output that we get by default, and expecting people to run it locally for full details? (There's also a middle-ground option where we parse out the name of the file but rather than parsing it and displaying it prettily, just dump the XML to stdout, so there's complete output in CI.)

blasten commented on script/tool/lib/src/lint_android_command.dart:
```diff
@@ -0,0 +1,56 @@
+
+    // TODO(stuartmorgan): Consider adding an XML parser to read and summarize
+    // all results. Currently, only the first three errors will be shown, and
+    // the rest are only in an XML file that won't be accessible for CI runs.
```
If moved to Github Actions, it could upload the generated HTML as an attachment using the GitHub API

stuartmorgan replied:
Looks like Cirrus could do it too: https://cirrus-ci.org/guide/writing-tasks/#artifacts-instruction. In fact: https://cirrus-ci.org/examples/#android-lint looks pretty interesting.

I can see how easy it will be to manage storage; we'd want them to auto-expire, ideally.

blasten commented on script/tool/lib/src/lint_android_command.dart:
```diff
@@ -0,0 +1,56 @@
+
+  @override
+  final String description = 'Runs "gradlew lint" on Android plugins.\n\n'
+      'Requires the example to have been build at least once before running.';
```
I think it only requires `flutter pub get`. Could this command invoke it before running ./gradlew?

blasten replied:
do you remember what file was missing?

stuartmorgan replied:
Yep, `gradlew`.

If you `git clean` a Flutter app (which removes `android/gradlew`), then `flutter pub get` or `flutter packages get` there's still no `android/gradlew`.

blasten commented on script/tool/lib/src/lint_android_command.dart:
```diff
@@ -0,0 +1,56 @@
+import 'common/package_looping_command.dart';
+import 'common/process_runner.dart';
+
+const String _gradleWrapper = 'gradlew';
```
If the idea is that it should eventually run on Windows, then it would also need to consider that this is gradlew.bat on Windows.

stuartmorgan replied:
I've been doing Windows support on an as-needed basis so far, but there's no reason I couldn't add it to this one out of the box. It would be useful for contributors.

blasten reviewed (approved):
LGTM other than the minor comments

blasten commented:
> Configure all plugins to not fail on error for now via the XML configuration, and remove that as they are fixed up?

`abortOnError false` in build.gradle  https://developer.android.com/studio/write/lint#gradle wdyt?

> Are there baseline options you want to set for all plugins before turning this on, or do you want to just make adjusting that part of fully opting each plugin into the run?

Ideally, plugins don't use baseline files since these are meant to hide existing issues. 

> Are you fine with punting on the XML parser for now, and just showing the partial output that we get by default, and expecting people to run it locally for full details? 

For the time being, could it dump the xml content to stdout?

stuartmorgan commented:
> > Configure all plugins to not fail on error for now via the XML configuration, and remove that as they are fixed up?
> 
> `abortOnError false` in build.gradle https://developer.android.com/studio/write/lint#gradle wdyt?
> 
> > Are there baseline options you want to set for all plugins before turning this on, or do you want to just make adjusting that part of fully opting each plugin into the run?
> 
> Ideally, plugins don't use baseline files since these are meant to hide existing issues.

Sorry, by baseline options I meant "are there any non-default lint settings we want to set". E.g., I don't know if there are things that aren't checked by default but we would want to enable.

But looking at baseline files more now that I know of them, that's strictly better than `abortOnError false` because we'll immediately start getting CI telling us about new failures for all plugins, rather than us having to go back and fix each existing plugin completely before getting any benefit at all for that plugin (since if regressions don't turn CI red we will, realistically, never notice them). So I'll see about setting those up as the interim solution, with an issue filed for removing them.

> > Are you fine with punting on the XML parser for now, and just showing the partial output that we get by default, and expecting people to run it locally for full details?
> 
> For the time being, could it dump the xml content to stdout?

I'll see how the parsed-by-Cirrus options pans out, and fall back to this if it doesn't give us what we need. Hopefully I'll have a new version of the PR up later today.

stuartmorgan commented:
@blasten Okay, I think this is ready for a full review now. Changes:
- Extracted a little helper file for Gradle commands (as I previously did with Xcode) and refactoring existing things to use it (as well as adding unit tests for it). It correctly handles the Windows vs non-Windows Gradle wrapper file name.
- Added "GradleDependency" to the ignore list in all plugins; it would be a regular source of out-of-band failures, which we absolutely don't want more of. (You might want to consider setting up a manual process of occasionally doing a trial PR that strips that out of everything and runs CI, to quickly get a list of everything that has dependencies that we might want to update.)
- Changed the lint command to only run lint for the plugin module itself, not transitively for everything, to avoid getting failures for things that aren't actually part of the plugin (like https://github.com/flutter/flutter/issues/88005)
  - I looked at adding `app` as well, but it seemed like the things it was flagging were likely not that interesting given that they are example apps. You may want to take a look as a follow-up and see if you do want to enable them.
- Added baseline files for plugins where necessary to make things green, and filed https://github.com/flutter/flutter/issues/88011 for cleaning them up
- Added the Cirrus command to upload, parse, and display the lint results as part of the run summary UI. Seems to work pretty well (you can't really see much now because of the exclusion files, but https://github.com/flutter/plugins/runs/3293114401 is an example from before I did that.

blasten commented:
> But looking at baseline files more now that I know of them, that's strictly better than abortOnError false

Sounds good


what is the awesome integration that shows linter warnings as GitHub comments in the PR?

blasten reviewed (approved):
LGTM. This is great!

stuartmorgan commented:
> what is the awesome integration that shows linter warnings as GitHub comments in the PR?

It's this bit:
https://github.com/flutter/plugins/pull/4206/files#diff-62587956f943bb2503db7bc6dd27d0d888074a1c0ecaab3f570ad611aff0f7bbR182-R186
The `format` entry causes it to parse and display the results.
